single_kernel_mongo/managers/tls.py

```diff
@@ -233,153 +175,235 @@ def enable_certificates_for_unit(self):
-        if not self.state.tls.get_secret(internal=False, label_name=SECRET_CERT_LABEL):
-            logger.debug("Waiting for external certificate.")
+
+        if self.state.client_tls_relation is not None and not self.is_certificate_available(
+            internal=False
+        ):
+            logger.debug("Waiting for client certificate.")
             return True
 
         return False
 
-    def update_tls_sans(self) -> None:
-        """Emits a certificate expiring event when sans in current certificates are out of date.
+    def get_tls_management_state(self) -> TlsManagementState:
+        """Pre-checks on TLS certificates management."""
+        if self.state.upgrade_in_progress:
+            return TlsManagementState.UPGRADE_IN_PROGRESS
+        if self.state.is_role(MongoDBRoles.MONGOS) and self.state.config_server_name is None:
+            return TlsManagementState.MONGOS_MISSING_CONFIG_SERVER
+        if not self.state.db_initialised:
+            if self.state.is_role(MongoDBRoles.MONGOS):
+                return TlsManagementState.MONGOS_DB_NOT_INITIALIZED
+            return TlsManagementState.DB_NOT_INTIALIZED
+        return TlsManagementState.EMPTY
+
+    def update_private_keys(self):
+        """Updates the private keys."""
+        peer_private_key = None
+        client_private_key = None
+
+        self.state.statuses.delete(
+            TLSStatuses.INVALID_PEER_PRIVATE_KEY.value, scope="unit", component=self.dependent.name
+        )
+        self.state.statuses.delete(
+            TLSStatuses.INVALID_CLIENT_PRIVATE_KEY.value,
+            scope="unit",
+            component=self.dependent.name,
+        )
 
-        This can occur for a variety of reasons:
-        1. Node port has been toggled on
-        2. Node port has been toggled off
-        3. The public K8s IP has changed
+        initial_peer_private_key = self.state.tls.peer_private_key
+        initial_client_private_key = self.state.tls.peer_private_key
 
-        Mongos k8s only.
-        """
-        for internal in [True, False]:
-            # if the certificate has already been requested, we do not want to re-request
-            # another one and lead to an infinite chain of certificate events.
-            if self.is_set_waiting_for_cert_to_update(internal):
-                continue
-            current_sans = self.get_current_sans(internal)
-            current_sans_ip = set(current_sans["sans_ips"]) if current_sans else set()
-            expected_sans_ip = set(self.get_new_sans()["sans_ips"]) if current_sans else set()
-            sans_ip_changed = current_sans_ip ^ expected_sans_ip
-
-            if not sans_ip_changed:
-                continue
+        if tls_peer_private_key_id := self.dependent.config.tls_peer_private_key_id:
+            if peer_private_key := self.update_private_key(tls_peer_private_key_id, internal=True):
+                self.dependent.tls_events.peer_certificate._private_key = peer_private_key
```

Comment:
Opened [this](https://github.com/canonical/tls-certificates-interface/issues/382) so I don't have to use private attribute to set the new key immediately.